portal-impl/src/com/liferay/portlet/documentlibrary/store/DLStoreImpl.java

```diff
@@ -109,11 +110,20 @@ public void copyFileVersion(
 			String fromVersionLabel, String toVersionLabel)
 		throws PortalException {
 
-		_wrappedStore.addFile(
-			companyId, repositoryId, fileName, toVersionLabel,
-			_getNullSafeInputStream(
-				_wrappedStore.getFileAsStream(
-					companyId, repositoryId, fileName, fromVersionLabel)));
+		if (_storeAreaProcessor != null) {
+			_storeAreaProcessor.copy(
+				StoreArea.LIVE.getPath(
+					companyId, repositoryId, fileName, fromVersionLabel),
+				StoreArea.LIVE.getPath(
+					companyId, repositoryId, fileName, toVersionLabel));
+		}
```

Comment:
It doesn't feel right to have this logic here. What about adding a `copyFile` method to Store interface with default naive implementation, and let implementing classes to make available an efficient copy if possible?

Reply:
That method existed in the past (along with other related methods -- e.g. `move`) and was removed on purpose from the API. Also, I created the processor interface just to avoid having to touch the main `Store` interface. It is best to keep all changes localized and separate from the `Store` interface as performance hooks (even if it is less "clean").